{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Trigger finger UKB genomic QC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Hail instance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import hail as hl\n",
    "import os\n",
    "from gnomad.utils.liftover import *\n",
    "from gnomad.utils.annotations import *\n",
    "from gnomad.sample_qc.pipeline import *\n",
    "from gnomad.sample_qc.ancestry import *\n",
    "\n",
    "tmp = \"/mnt/grid/janowitz/home/skleeman/tmp2\"\n",
    "os.environ[\"SPARK_LOCAL_DIRS\"]=tmp\n",
    "\n",
    "os.environ[\"PYSPARK_SUBMIT_ARGS\"] =\"--driver-memory 200g --executor-memory 2g pyspark-shell\"\n",
    "\n",
    "hl.init(default_reference='GRCh38', master='local[16]',min_block_size=128, local_tmpdir=tmp, tmp_dir=tmp)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge UK biobank PLINK data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Shell script\n",
    "\n",
    "plink \\\n",
    "  --bed /mnt/grid/ukbiobank/data/genotypes/plink/ukb_cal_chr1_v2.bed \\\n",
    "  --bim /mnt/grid/janowitz/home/skleeman/ukbiobank/rawdata/ukb_snp_chr1_v2.bim \\\n",
    "  --fam /mnt/grid/ukbiobank/data/ApplicationXXXXX/rawdata/ukbXXXXX_cal_chr1_v2_s488250.fam \\\n",
    "  --merge-list ~/list_beds.txt \\\n",
    "  --make-bed --out ukb_merged \\"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed = '/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/merged_plink/ukb_merged.bed'\n",
    "bim = '/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/merged_plink/ukb_merged.bim'\n",
    "fam = '/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/merged_plink/ukb_merged.fam'\n",
    "\n",
    "ukb = hl.import_plink(bed = bed, bim = bim, fam = fam, reference_genome='GRCh37',\n",
    "                     min_partitions = 150)\n",
    "\n",
    "ukb.write('/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/ukb_merged_raw.mt', overwrite=True) #Save raw data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## UKB pre-process"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter - GWAS\n",
    "\n",
    "MAF > 1% (I checked none are >99%), missing rate < 5%\n",
    "\n",
    "This will form the input to BOLT-LMM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import UKB data in PLINK form\n",
    "\n",
    "ukb = hl.read_matrix_table('/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/ukb_merged_raw.mt')\n",
    "\n",
    "print(ukb.count()) #How many variants to start with?\n",
    "\n",
    "#Filtering - based partly on https://github.com/Nealelab/UK_Biobank_GWAS/blob/master/imputed-v2-gwas/README.md\n",
    "\n",
    "ukb = hl.variant_qc(ukb) #Default Hail variant QC pipeline\n",
    "ukb = ukb.filter_rows(ukb.variant_qc.AF[1] > 0.01) #MAF > 1%\n",
    "ukb = ukb.filter_rows(ukb.variant_qc.call_rate > 0.95) #Filter by call rate > 95%\n",
    "\n",
    "print(ukb.count()) #How many variants are left?\n",
    "\n",
    "ukb.write('/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/ukb_grch37_filtered_forgwas.mt', overwrite=True) #Save pruned MT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ukb = hl.read_matrix_table('/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/ukb_merged_raw.mt')\n",
    "\n",
    "print(ukb.count()) #How many variants to start with?\n",
    "\n",
    "#Filtering - based partly on https://github.com/Nealelab/UK_Biobank_GWAS/blob/master/imputed-v2-gwas/README.md\n",
    "\n",
    "ukb = hl.variant_qc(ukb) #Default Hail variant QC pipeline\n",
    "ukb = ukb.filter_rows(ukb.variant_qc.AF[1] > 0.99) #MAF > 1%\n",
    "\n",
    "print(ukb.count()) #How many variants are left?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ancestry analysis, sample QC\n",
    "\n",
    "### Merge UKB and Reference cohorts\n",
    "\n",
    "Biallelic SNPs, lift to grch38, remove allele mismatches, set AF > 0.1% and call rate >99% in UKB, identify high quality reference variants in WGS data (call rate >95%, biallelic, SNPs cf. indels)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import hail as hl\n",
    "import os\n",
    "from gnomad.utils.liftover import *\n",
    "from gnomad.utils.annotations import *\n",
    "from gnomad.sample_qc.pipeline import *\n",
    "import subprocess\n",
    "\n",
    "#Define memory and CPU availability\n",
    "\n",
    "tmp = \"/mnt/grid/janowitz/home/skleeman/tmp\"\n",
    "\n",
    "os.environ[\"SPARK_LOCAL_DIRS\"]=tmp\n",
    "os.environ[\"PYSPARK_SUBMIT_ARGS\"] =\"--master local[96] --driver-memory 960G pyspark-shell\"\n",
    "\n",
    "\n",
    "hl.init(default_reference='GRCh38', master ='local[96]', local='local[96]',min_block_size=128, local_tmpdir=tmp, tmp_dir=tmp)\n",
    "\n",
    "\n",
    "#Import UKB\n",
    "ukb = hl.read_matrix_table('/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/ukb_merged_raw.mt')\n",
    "print(ukb.count()) #How many variants to start with?\n",
    "ukb = default_lift_data(ukb) #GNOMAD pipeline for liftover, including reverse complement on negative strand\n",
    "\n",
    "#Filtering\n",
    "\n",
    "ukb = hl.variant_qc(ukb) #Default Hail variant QC pipeline\n",
    "ukb = ukb.filter_rows(hl.len(ukb.alleles) == 2) #Biallelic SNPs only\n",
    "ukb = ukb.filter_rows(ukb.ref_allele_mismatch == False) #Remove alleles with reference mismatch ('allele flips')\n",
    "ukb = ukb.filter_rows(ukb.variant_qc.AF[1] > 0.001) #MAF > 0.1%\n",
    "ukb = ukb.filter_rows(ukb.variant_qc.call_rate > 0.99) #Filter by call rate > 99%\n",
    "ukb = ukb.checkpoint('/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/ukb_grch38_filtered.mt', overwrite=True) #Save filtered MT\n",
    "\n",
    "#Exclude LD intervals from plinkQC package, LD pruning in PLINK (not working in Hail due to bug)\n",
    "intervals = hl.import_bed('/mnt/grid/janowitz/home/skleeman/ukbiobank/cancergwas/remove_ld_grch38.bed',\n",
    "                         reference_genome='GRCh38')\n",
    "ukb = ukb.filter_rows(hl.is_defined(intervals[ukb.locus]),keep=False)\n",
    "\n",
    "hl.export_plink(ukb, '/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/pre_ld_plink',\n",
    "                fam_id = ukb.fam_id, ind_id = ukb.s, pat_id = ukb.pat_id, mat_id = ukb.mat_id,\n",
    "                is_female = ukb.is_female)\n",
    "\n",
    "commands = '''\n",
    "plink --bfile /mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/pre_ld_plink \\\n",
    "--indep-pairwise 50 5 0.2 --out /mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/prune --threads 96 --memory 50000 \\\n",
    "--allow-extra-chr\n",
    "\n",
    "plink --bfile /mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/pre_ld_plink \\\n",
    "--extract /mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/prune.prune.in \\\n",
    "--make-bed --out /mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/post_ld_plink --threads 96 --memory 50000 \\\n",
    "--allow-extra-chr\n",
    "'''\n",
    "\n",
    "output = subprocess.check_output(commands, shell=True)\n",
    "print(output)\n",
    "\n",
    "bed = '/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/post_ld_plink.bed'\n",
    "bim = '/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/post_ld_plink.bim'\n",
    "fam = '/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/post_ld_plink.fam'\n",
    "\n",
    "ukb = hl.import_plink(bed = bed, bim = bim, fam = fam, reference_genome='GRCh38',\n",
    "                     min_partitions = 150)\n",
    "print(ukb.count())\n",
    "ukb = ukb.checkpoint('/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/ukb_grch38_pruned.mt', overwrite=True) #Save filtered MT\n",
    "\n",
    "#Import REF\n",
    "this_ref = hl.read_matrix_table('/mnt/grid/janowitz/rdata_norepl/ref.mt') #Dense form (!!!)\n",
    "\n",
    "this_ref = hl.variant_qc(this_ref) #Default Hail variant QC pipeline\n",
    "this_ref = this_ref.filter_rows((hl.len(this_ref.alleles) == 2) & hl.is_snp(this_ref.alleles[0], this_ref.alleles[1]))\n",
    "this_ref = this_ref.filter_rows(this_ref.variant_qc.call_rate > 0.95) #Filter by call rate > 95%\n",
    "this_ref = this_ref.naive_coalesce(250)\n",
    "\n",
    "this_ref = this_ref.checkpoint('/mnt/grid/janowitz/home/references/1k_hgdp/ref_gnomadfilters.mt', overwrite=True) #Save filtered MT\n",
    "\n",
    "#Intersect\n",
    "ukbb_in_ref = ukb.filter_rows(hl.is_defined(this_ref.rows()[ukb.row_key]))\n",
    "print('sites in ref and UKBB data, inds in UKBB: ' + str(ukbb_in_ref.count()))\n",
    "\n",
    "ref_in_ukbb = this_ref.filter_rows(hl.is_defined(ukb.rows()[this_ref.row_key]))\n",
    "print('sites in ref and UKBB data, inds in ref: ' + str(ref_in_ukbb.count()))\n",
    "\n",
    "#Export\n",
    "ref_in_ukbb.write('/mnt/grid/janowitz/home/references/1k_hgdp/ref_intersect.mt', overwrite=True)\n",
    "ukbb_in_ref.write('/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/ukb_grch38_intersect.mt',overwrite=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "plinkQC reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "this_ref = hl.read_matrix_table('/mnt/grid/janowitz/home/references/1k_hgdp/ref_intersect.mt')\n",
    "this_ref = default_lift_data(this_ref) \n",
    "\n",
    "related_samples_to_remove_ref = hl.read_table(\"/mnt/grid/janowitz/home/references/1k_hgdp/related_remove_ref.ht\")\n",
    "\n",
    "this_ref = this_ref.filter_cols(hl.is_defined(related_samples_to_remove_ref[this_ref.col_key]), keep=False)\n",
    "\n",
    "this_ref.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "this_ref = hl.read_matrix_table('/mnt/grid/janowitz/home/references/1k_hgdp/ref_intersect.mt') #Dense form (!!!)\n",
    "snps = this_ref.rows()\n",
    "snps = snps.select_globals()\n",
    "snps = snps.select()\n",
    "snps.write('/mnt/grid/janowitz/home/references/1k_hgdp/ukb_snp_reference.ht',overwrite=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run PCA and RF classifier\n",
    "\n",
    "Train classifier on reference set, pre-process with removal of related samples to 3rd degree (PC-Relate threshold of 0.05), extract 10 principal components, train RF classifier using populations labels subdivided per https://docs.google.com/spreadsheets/d/1jenSz_HnbA1kBESaUmur3Ob72-EPXJgfUWhbz5UdltA/edit#gid=433808438. Define populations using RF classifer probability threshold of 0.70.\n",
    "\n",
    "Run on cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import pandas as pd\n",
    "from gnomad.sample_qc.ancestry import *\n",
    "\n",
    "\n",
    "\n",
    "this_ref = hl.read_matrix_table('/mnt/grid/janowitz/home/references/1k_hgdp/ref_intersect.mt')\n",
    "ukb = hl.read_matrix_table('/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/ukb_grch38_intersect.mt')\n",
    "\n",
    "\n",
    "#Find related individuals in 1000K/HGDP set\n",
    "\n",
    "relatedness_ht = hl.pc_relate(this_ref.GT, 0.01, k=10, min_kinship=0.05, block_size=512)\n",
    "\n",
    "related_samples_to_remove_ref = hl.maximal_independent_set(relatedness_ht.i, relatedness_ht.j, False)\n",
    "\n",
    "\n",
    "#Run PCA\n",
    "\n",
    "#--> Reference, label with inferred populations, exclude relateds\n",
    "_, scores_pca_ref, loadings_pca_ref = run_pca_with_relateds(this_ref, related_samples_to_remove_ref, \n",
    "                                                               n_pcs=10, autosomes_only=True)\n",
    "\n",
    "#--> UKB\n",
    "\n",
    "scores_pca_ukb = ancestry.pc_project(mt = ukb, loadings_ht = loadings_pca_ref)\n",
    "\n",
    "scores_pca_ref.write(\"/mnt/grid/janowitz/home/references/1k_hgdp/scores_pca_ref.ht\", overwrite=True)\n",
    "scores_pca_ukb.write(\"/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/scores_pca_ukb.ht\", overwrite=True)\n",
    "related_samples_to_remove_ref.write(\"/mnt/grid/janowitz/home/references/1k_hgdp/related_remove_ref.ht\",overwrite=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run on Jupyter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "scores_pca_ref = hl.read_table(\"/mnt/grid/janowitz/home/references/1k_hgdp/scores_pca_ref.ht\")\n",
    "scores_pca_ukb = hl.read_table(\"/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/scores_pca_ukb.ht\")\n",
    "this_ref = hl.read_matrix_table('/mnt/grid/janowitz/home/references/1k_hgdp/ref_intersect.mt')\n",
    "\n",
    "merge = scores_pca_ref.union(scores_pca_ukb)\n",
    "\n",
    "merge = merge.annotate(\n",
    "    training_pop=this_ref.cols()[merge.key].labeled_subpop)\n",
    "\n",
    "recode = pd.read_excel('/mnt/grid/janowitz/home/references/1k_hgdp/recode.xlsx')\n",
    "recode_ht = hl.Table.from_pandas(recode, key='labeled_subpop')\n",
    "\n",
    "merge = merge.annotate(\n",
    "    training_pop=recode_ht[merge.training_pop].superpop)\n",
    "\n",
    "predictions_ref, classifer_rf_ref = assign_population_pcs(merge, pc_cols = merge.scores, known_col = 'training_pop', seed=501, min_prob = 0.70, missing_label='Other')\n",
    "\n",
    "ukb_predictions = predictions_ref.semi_join(scores_pca_ukb) #Subset UKB samples\n",
    "\n",
    "ukb_predictions.write(\"/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/ukb_predictions.ht\", overwrite=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Summarize population calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ukb_predictions_pd = ukb_predictions.to_pandas()\n",
    "ukb_predictions_pd = ukb_predictions_pd[[\"s\", \"pop\"]]\n",
    "ukb_predictions_pd = ukb_predictions_pd[ukb_predictions_pd['pop'] != \"Other\"]\n",
    "ukb_predictions_pd['pop'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Trigger finger GWAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import hail as hl\n",
    "import os\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "from gnomad.sample_qc.ancestry import *\n",
    "\n",
    "#Define memory and CPU availability\n",
    "\n",
    "tmp = \"/mnt/grid/janowitz/rdata_norepl/tmp\"\n",
    "\n",
    "os.environ[\"SPARK_LOCAL_DIRS\"]=tmp\n",
    "os.environ[\"PYSPARK_SUBMIT_ARGS\"] =\"--conf spark.network.timeout=15m --conf spark.executor.heartbeatInterval=10m --conf spark.memory.fraction=1.0 --driver-memory 2880G pyspark-shell\"\n",
    "\n",
    "\n",
    "hl.init(default_reference='GRCh37', master ='local[96]',min_block_size=128, local_tmpdir=tmp, tmp_dir=tmp)\n",
    "\n",
    "\n",
    "#Pre-process ancestry predictions, merge with phenotype data\n",
    "ukb_predictions = hl.read_table(\"/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/ukb_predictions.ht\")\n",
    "\n",
    "ukb_predictions_pd = ukb_predictions.to_pandas()\n",
    "ukb_predictions_pd = ukb_predictions_pd[[\"s\", \"pop\"]]\n",
    "ukb_predictions_pd = ukb_predictions_pd[ukb_predictions_pd['pop'] != \"Other\"]\n",
    "print(ukb_predictions_pd['pop'].value_counts(), flush=True)\n",
    "\n",
    "phenotypes = pd.read_csv('/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/trigger_gwas.csv')\n",
    "phenotypes = phenotypes.drop(phenotypes.columns[0], axis=1)\n",
    "phenotypes = phenotypes.rename(columns={'id': 's'})\n",
    "phenotypes['s'] = phenotypes.s.astype(int)\n",
    "ukb_predictions_pd['s'] = ukb_predictions_pd.s.astype(int)\n",
    "\n",
    "phenotypes = phenotypes.merge(ukb_predictions_pd, how='inner', on='s')\n",
    "\n",
    "populations = phenotypes['pop'].unique().tolist()\n",
    "#populations = [\"CSA\"]\n",
    "\n",
    "#Add relationship matrix (UKB-provided KING analysis)\n",
    "rel = hl.import_table('/mnt/grid/ukbiobank/data/ApplicationXXXXX/rawdata/ukbXXXXX_rel_s488250.dat', impute=True,delimiter='\\s+', \n",
    "                     types={'ID1': hl.tstr, 'ID2': hl.tstr})\n",
    "rel = rel.select(rel.ID1, rel.ID2, rel.Kinship)\n",
    "\n",
    "#Import GRCh37 data (filtered to call rate, MAF, no H-W filtering at present)\n",
    "ukb_gwas = hl.read_matrix_table('/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/ukb_grch37_filtered_forgwas.mt')\n",
    "ukb_gwas = ukb_gwas.filter_rows(ukb_gwas.variant_qc.AC[1] > 100) #As per Regenie documentation\n",
    "print(ukb_gwas.count(), flush=True)\n",
    "\n",
    "#Prune GRCh37 data using SNP list (in_PCA) from marker paper\n",
    "snp_ukb = hl.import_table('/mnt/grid/janowitz/home/skleeman/ukbiobank/cancergwas/relatedness/ukb_snp_qc.txt', impute=True, delimiter='\\s+')\n",
    "snp_ukb = snp_ukb.filter(snp_ukb.in_PCA==1)\n",
    "snp_ukb = snp_ukb.key_by(snp_ukb.rs_id)\n",
    "ukb_gwas_pruned = ukb_gwas.filter_rows(hl.is_defined(snp_ukb[ukb_gwas.rsid]))\n",
    "\n",
    "for pop in populations:\n",
    "    print(pop, flush=True)\n",
    "    phenotypes['s'] = phenotypes.s.astype(str)\n",
    "    \n",
    "    #Subset covariate file, make suitable for BOLT-LMM\n",
    "    \n",
    "    subset = phenotypes[phenotypes['pop'] == pop]\n",
    "    \n",
    "    frame = { 'FID': subset['s'], 'IID': subset['s'], 'year_of_birth': subset['year_of_birth_f34_0_0'],\n",
    "             'sex': subset['sex_f31_0_0'],'array': subset['genotype_measurement_batch_f22000_0_0'],\n",
    "        'centre': subset['uk_biobank_assessment_centre_f54_0_0']}\n",
    "    frame2 = { 'FID': subset['s'], 'IID': subset['s'], 'base_analysis':subset['base_analysis'], 'base_analysis_and_self':subset['base_analysis_and_self'], 'double_positive':subset['double_positive'],'base_without_CTS':subset['base_without_CTS'],'mixed_tf_cts':subset['mixed_tf_cts']}\n",
    "    \n",
    "    covariates = pd.DataFrame(frame) \n",
    "    phenotypes_out = pd.DataFrame(frame2) \n",
    "    print(phenotypes_out['trigger_scar'].value_counts(), flush=True)\n",
    "    print(phenotypes_out['hypertrophic_scar'].value_counts(), flush=True)\n",
    "    \n",
    "    #Subset GRCH37 version of UK biobank dataset\n",
    "    \n",
    "    subset_ht = hl.Table.from_pandas(subset['s'].to_frame(), key='s') \n",
    "    ukb_pop_gwas = ukb_gwas.filter_cols(hl.is_defined(subset_ht[ukb_gwas.col_key]))\n",
    "    ukb_pop_gwas = hl.variant_qc(ukb_pop_gwas)\n",
    "    ukb_pop_gwas = ukb_pop_gwas.filter_rows(ukb_pop_gwas.variant_qc.AC[1] > 20) #As per Regenie documentation\n",
    "    ukb_pop_gwas_pruned = ukb_gwas_pruned.filter_cols(hl.is_defined(subset_ht[ukb_gwas_pruned.col_key]))\n",
    "\n",
    "    #Save filtered PLINK file (GRCH37, include rsid)\n",
    "    \n",
    "    folder = '/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/gwas_trigger/' + pop + '/filtered_grch37'\n",
    "    \n",
    "    if not os.path.exists(folder + '.bim'):\n",
    "        hl.export_plink(ukb_pop_gwas, folder,\n",
    "                    fam_id = ukb_pop_gwas.fam_id, ind_id = ukb_pop_gwas.s, pat_id = ukb_pop_gwas.pat_id, mat_id = ukb_pop_gwas.mat_id,\n",
    "                    is_female = ukb_pop_gwas.is_female, varid = ukb_pop_gwas.rsid, pheno = -9)\n",
    "    \n",
    "    #Run PCA on all filtered variants, per population, removing relateds subjects per population\n",
    "    \n",
    "    rel_sub = rel.filter(hl.is_defined(subset_ht[rel.ID1]) & hl.is_defined(subset_ht[rel.ID2]))\n",
    "    pairs = rel_sub.filter(rel_sub['Kinship'] > 0.0442) #Third degree relatives per KING documentation\n",
    "    related_samples_to_remove = hl.maximal_independent_set(pairs.ID1, pairs.ID2, False)\n",
    "    \n",
    "    print(\"Related + \", pop, related_samples_to_remove.count(), flush=True)\n",
    "    \n",
    "    _, scores_pca_ref, _ = run_pca_with_relateds(ukb_pop_gwas_pruned, related_samples_to_remove, \n",
    "                                                               n_pcs=10, autosomes_only=True)\n",
    "    \n",
    "    scores_pca_ref = scores_pca_ref.transmute(**{f'PC{i}': scores_pca_ref.scores[i - 1] for i in range(1, 11)})\n",
    "    scores_pd = scores_pca_ref.to_pandas()\n",
    "    scores_pd = scores_pd.rename(columns={'s': 'IID'})\n",
    "    \n",
    "    #Add PCA results to covariates file, then save\n",
    "    \n",
    "    covariates = covariates.merge(scores_pd, left_on='IID', right_on='IID')\n",
    "    folder = '/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/gwas_trigger/' + pop + '/covariates.tsv'\n",
    "    covariates.to_csv(folder, sep=\"\\t\", index=False)\n",
    "    folder = '/mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/gwas_trigger/' + pop + '/phenotypes.tsv'\n",
    "    phenotypes_out.to_csv(folder, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "for folder in /mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/gwas_trigger/*/; do\n",
    "\n",
    "base=$(basename $folder)\n",
    "\n",
    "echo $base\n",
    "\n",
    "cat <<EOF >./scripts/$base.sh\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -pe threads 16\n",
    "#$ -l m_mem_free=8G\n",
    "#$ -N gwas_$base\n",
    "#$ -o $base.txt\n",
    "#$ -e $base.txt\n",
    "\n",
    "cd /mnt/grid/ukbiobank/data/ApplicationXXXXX/skleeman/gwas_trigger/$base\n",
    "\n",
    "regenie \\\n",
    "  --step 1 --threads 16 \\\n",
    "  --bed filtered_grch37 \\\n",
    "  --phenoFile phenotypes.tsv \\\n",
    "  --covarFile covariates.tsv \\\n",
    "  --bt \\\n",
    "  --bsize 1000 \\\n",
    "  --lowmem \\\n",
    "  --lowmem-prefix /mnt/grid/janowitz/rdata_norepl/tmp/regenie/$base \\\n",
    "  --out ukb_step1_BT\n",
    "\n",
    "for chr in {1..22}\n",
    "do\n",
    "\n",
    "regenie \\\n",
    "  --step 2 --threads 16 \\\n",
    "  --bgen /mnt/grid/ukbiobank/data/ApplicationXXXXX/imputed_missing/ukb_imp_chr\\${chr}_v3.bgen \\\n",
    "  --sample /mnt/grid/ukbiobank/data/ApplicationXXXXX/rawdata/ukbXXXXX_imp_chr1_v3_s487282.sample \\\n",
    "  --phenoFile ukb_phenotypes_BT.txt \\\n",
    "  --covarFile ukb_covariates.txt \\\n",
    "  --bt \\\n",
    "  --firth 0.01 --approx \\\n",
    "  --pred ukb_step1_BT_pred.list \\\n",
    "  --bsize 400 \\\n",
    "  --minINFO 0.8 \\\n",
    "  --split \\\n",
    "  --out ukb_step2_BT_chr\\${chr}\n",
    "\n",
    "done\n",
    "\n",
    "EOF\n",
    "done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python/3.7.4 (src)",
   "language": "python",
   "name": "python_3.7.4-gcccore-8.3.0"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
